Disable context menu on resource url for download

vertein opened the pull request:
We have a button that does work masquerading as a link to a file.  This results in an issue with uPortal's naming convention of URL's.  Example from what the URL is on this link: `http://localhost:8080/portal/f/u28l1s4/p/PayrollInformation.n94/normal/irs_statement.pdf.resource.uP?pP_docId=3`.  Normally we can add some info in the context header telling the browser to expect a file returned and file name to save it as.  Safari does not honor this when saving from the context menu (menu that appears when a user right clicks something).
This commit adds a class that can be used to disable the context menu for certain elements.
`class="contextMenuDisallowed"`

This class is added to the link shown below.  It will prevent users in Safari from bring up the context menu and clicking save-as.
![image](https://cloud.githubusercontent.com/assets/5521429/2842782/45b760e2-d078-11e3-80c8-ca86f4e6b094.png)

Browser detection and functionality tested with Mac: Safari, Chrome, Firefox. Linux: Chrome, Firefox. Windows: Chrome, Firefox, Internet Explorer.

vertein committed:
Disable context menu on resource url for download

Safari does not honor context headers for
download names.
This commit adds a class that can be used to
disable the context menu for certain elements.
The user of course can still command + click the
link and normal click a link.

apetro commented on hrs-portlets-webapp/src/main/webapp/js/portlet.js:
```diff
@@ -0,0 +1,31 @@
+
+/*
+ * Switch libraries to extreme noConflict mode, keeping a reference to it in the
+ * blackboardPortlet variable if one doesn't already exist
```
Stray `blackboardPortlet` reference; also another one on line 19.

vertein replied:
lol.  You're right.  I'll get that changed.

apetro commented:
I like this solution a lot.  It's tightly tailored to the one offending browser, it prevents access to the broken browser feature.

Will Safari users "get" that they can just click on the link to get the file, or will they frustratedly right-click to no effect and curse us?  I imagine they'll give up and just click, but I guess I don't know that without user testing.  I'm fine with promoting this fix up into Test etc. and seeing how Product Owners feel, whether they want to put it in Prod and See What Happens.  Or, it might be more usable with a tooltip (Safari-specific tooltip?) to tell users that the context menu is disabled in Safari to work around a browser bug?

apetro commented on hrs-portlets-webapp/src/main/webapp/WEB-INF/jsp/payrollInformation.jsp:
```diff
@@ -89,7 +89,7 @@
             </thead>
             <tbody>
-                <tr rsf:id="row:" class="dl-clickable">
+                <tr rsf:id="row:" class="dl-clickable contextMenuDisallowed">
```
Nitpick: on skim, `contextMenuDisallowed` suggests to me that the context menu is, well, disallowed, whereas in the details this is really only `contextMenuDisallowedOnSafari`.  I'm not sure what to do with that thought -- thinking it, rolling eyes, and just moving on is a fine option.  But maybe it gets to just what this class is meant to flag.

Thought: should the context-menu-disallowing behavior actually be applied to the `dl-clickable` class?  That is, what we're saying about this element is that it's a click-for-download element, and then what we're saying in the JavaScript is that, hey, Safari is screwed up for _all_ click-to-download (`dl-clickable`) elements.  It's not so much a characteristic of this element that it's context-menu-disallowed as it's a characteristic of this element that it's a `dl-clickable` and it's a characteristic of Safari, expressed in that event handler on line ~142, that there's special handling of context menu invocation needed for Safari consumption of click-to-download elements, namely, suppress the menu in that context.

apetro commented on hrs-portlets-webapp/src/main/webapp/WEB-INF/jsp/payrollInformation.jsp:
```diff
@@ -133,24 +133,36 @@
+        
+        <%-- Disable Right Clicks in Safari browsers --%>
+        <%-- Safari (not webkit) does not honor context header for naming downloads--%>
+        if (Object.prototype.toString.call(window.HTMLElement).indexOf('Constructor') > 0) {
```
Is this fanciness meant to detect Safari?  Is that worth documenting more explicitly, or even externalizing into a `usingSafariBrowser` or `usingBrowserWithDefectiveContextMenuDownloadOption` function or variable?

timlevett commented on hrs-portlets-webapp/src/main/webapp/WEB-INF/jsp/payrollInformation.jsp:
```diff
@@ -225,7 +237,8 @@
+        });
     });    
-})(dl_v1.jQuery, dl_v1.fluid, dl_v1);
+})(dl_v1.jQuery, dl_v1.fluid, dl_v1, hrsPortlet.jQuery);
```
two versions of jquery?

vertein replied:
I'll try to remove the first one and see what happens.

timlevett commented on hrs-portlets-webapp/src/main/webapp/WEB-INF/jsp/payrollInformation.jsp:
```diff
@@ -133,24 +133,36 @@
-(function($, fluid, dl) {
-    $(function() {
+(function($, fluid, dl, jQuery) {
+    var $ = hrsPortlet.jQuery;
```
$ is already jquery via the function declaration.

vertein replied:
This is exactly the same thing we had before.  It doesn't work without it.  We had this discussion on a marketplace javascript file.

timlevett commented on hrs-portlets-webapp/src/main/webapp/resources.xml:
```diff
@@ -23,9 +23,14 @@
   <css>css/HRSPortlet.css</css>
+  
+  <js included="plain" resource="true">/rs/jquery/1.11.0/jquery-1.11.0.js</js>
+  <js included="aggregated" resource="true">/rs/jquery/1.11.0/jquery-1.11.0.min.js</js>
```
Surprised that jQuery is not already in this.

vertein commented:
Removed the entire file that reference Blackboard.
Talked face to face about whether Safari users will `get it`.  Agreement is that we feel something should be done, but unsure what.  Issue tracked in trello for UX consultant and out of scope for this PR.
Removed `contextMenuDisallowed` class.  I like that suggestion.
I'm not going to name the if-statement condition, but I did add more clarification to the comment.
Addressed all of the jquery comments by removing the second set of jquery.

I think all the comments have been addressed.  Let me know if I missed anything, or if revisiting an idea is worth it.

timlevett commented:
:+1: Looks good to me.
